File: android/app/src/main/java/com/devicelens/app/helpers/RelativeTimeFormatter.kt
package com.devicelens.app.helpers

import java.util.concurrent.TimeUnit
import javax.inject.Inject

class RelativeTimeFormatter @Inject constructor() {

    fun format(epochMs: Long): String {
        val now = System.currentTimeMillis()
        val diff = now - epochMs

        return when {
            diff < TimeUnit.MINUTES.toMillis(1) -> "just now"
            diff < TimeUnit.HOURS.toMillis(1) -> {
                val mins = TimeUnit.MILLISECONDS.toMinutes(diff)
                if (mins == 1L) "1 minute ago" else "$mins minutes ago"
            }
            diff < TimeUnit.DAYS.toMillis(1) -> {
                val hours = TimeUnit.MILLISECONDS.toHours(diff)
                if (hours == 1L) "1 hour ago" else "$hours hours ago"
            }
            diff < TimeUnit.DAYS.toMillis(2) -> "yesterday"
            diff < TimeUnit.DAYS.toMillis(7) -> {
                val days = TimeUnit.MILLISECONDS.toDays(diff)
                "$days days ago"
            }
            diff < TimeUnit.DAYS.toMillis(30) -> {
                val weeks = TimeUnit.MILLISECONDS.toDays(diff) / 7
                if (weeks == 1L) "1 week ago" else "$weeks weeks ago"
            }
            else -> {
                val months = TimeUnit.MILLISECONDS.toDays(diff) / 30
                if (months == 1L) "1 month ago" else "$months months ago"
            }
        }
    }
}

File: ios/DeviceLens/Helpers/RelativeTimeFormatter.swift
import Foundation

struct RelativeTimeFormatter {
    func format(_ date: Date) -> String {
        let diff = Date().timeIntervalSince(date)

        switch diff {
        case ..<60:
            return "just now"
        case ..<3600:
            let mins = Int(diff / 60)
            return mins == 1 ? "1 minute ago" : "\(mins) minutes ago"
        case ..<86400:
            let hours = Int(diff / 3600)
            return hours == 1 ? "1 hour ago" : "\(hours) hours ago"
        case ..<172800:
            return "yesterday"
        case ..<604800:
            let days = Int(diff / 86400)
            return "\(days) days ago"
        case ..<2592000:
            let weeks = Int(diff / 604800)
            return weeks == 1 ? "1 week ago" : "\(weeks) weeks ago"
        default:
            let months = Int(diff / 2592000)
            return months == 1 ? "1 month ago" : "\(months) months ago"
        }
    }
}
